docs/en/04_Changelogs/4.0.0.md

```diff
@@ -1095,6 +1095,55 @@ Some methods on `Requirements` have had their method signatures changed:
 A new config `Requirements_Backend.combine_in_dev` has been added in order to allow combined files to be
 forced on during development. If this is off, combined files is only enabled in live environments.
 
+Form validation has been refactored significantly. A new `FormMessage` trait has been created to
+handle field-level and form-level messages. This has the following properties:
+
+* `setMessage` to assign a message, type, and cast
+* `getMessage` retrieves the message string
+* `getMessageType` retrieves the message type (E.g. error, good, info)
+* `getMessageCast` retrieves the cast type
+* `getMessageCastingHelper` retrieves the DBField cast to use for the appropriate message cast
+* `getSchemaMessage` encodes this message for form schema use in ReactJS.
+
+`Form` methods have been changed:
+
+* `validate` method now returns a `ValidationResult` instance. This is no longer automatically
+  persisted in the state by default, unless a redirection occurs. You can also save any response in the
+  state by manually invoking `saveFormState` inside a custom validation response handler.
+* `setupFormErrors` renamed to `restoreFormState`
+* `resetValidation` renamed to `clearFormState`
+* `loadMessagesFrom` method created to load a ValidationResult into a form.
+* `setMessage`. third parameter is now $cast type
+* `messageForForm`. removed
+* `getSessionValidationResult` / `setSessionValidationResult` used to get / set session errors
+* `getSessionData` / `setSessionData` used to get / set field values cached in the session
+* `getAjaxErrorResponse` and `getRedirectReferer` created to simplify `getValidationErrorResponse`
```

Comment:
Inline methods increases code complexity (of the combined block) and reduces readability. These methods are in fact not temporary; All that will change is that they will be moved to FormResponseHandler, not removed.